### PR TITLE
[v3] add time/duration parsing support

### PR DESCRIPTION
**Pull request by kukasaurs**
**What's Changed**

- Added support for parsing two types: `time.Time` and `time.Duration`.
- Improved tests in `env_time_test.go` to cover edge cases and ensure correctness.

### CODE REVIEW

**Comment by pugnack** [comment]
title => [v3] add time/duration parsing support

т.к. это не фикс, а добавление функционала

**Comment by pugnack** [comment]
description => 

* не имеет глубокого смысла писать о том, что ты что-то локально тестировала и у тебя тесты локально проходят — это само собой разумеющийся пункт без которого и смысла нет создавать PR.
* Pull Request замени на What's Changed, т.к. и так понятно, что это Pull Request. Example можно глянуть [тут](https://github.com/unistack-org/micro-client-http/pull/156). Ну и шрифты крупные лучше не использовать на такие краткие описания, bold достаточно.
* не надо упоминать номер таски из проекта, т.к. это внутренняя инфа, которая не должна утекать наружу, в задаче лучше указать ссылку на PR на gihub

итого:
* ты добавила саппорт двух типов (time.Time, time.Duration)
* улучшила тесты

File: env_time_test.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"os"
+	"testing"
+	"time"
+
+	"go.unistack.org/micro/v3/config"
+
+	env "go.unistack.org/micro-config-env/v3"
+)
+
+type EnvConfig struct {
```
Comment (pugnack) [inline env_time_test.go]: Лучше спрятать объявление типа в сам тест, чтобы не плодить публичных типов на уровне пакета, т.к. он нужен только конкретному тесту.

File: env_time_test.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	IntValue      int           `env:"INT_VALUE"`
+	DurationValue time.Duration `env:"CACHE_TTL"`
+	TimeValue     time.Time     `env:"START_DATE"`
+}
+
+func TestEnv_BasicTypes(t *testing.T) {
+	cases := []struct {
```
Comment (pugnack) [inline env_time_test.go]: rename: cases => tests

File: env_time_test.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		name    string
+		envVar  string
+		envVal  string
+		checker func(t *testing.T, cfg *EnvConfig)
+	}{
```
Comment (pugnack) [inline env_time_test.go]: договаривались, что ты напишешь качественные тесты, а не 2-3 типа кейса покроешь, переименуй тест в TestEnv_SupportedTypes и определи в тест-кейсах все типы, которые мы поддерживаем

также не имеет смысле держать отдельный файл под это, перенеси в env_tests.go

File: env_time_test.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	DurationValue time.Duration `env:"CACHE_TTL"`
+	TimeValue     time.Time     `env:"START_DATE"`
+}
+
+func TestEnv_BasicTypes(t *testing.T) {
+	cases := []struct {
+		name    string
+		envVar  string
+		envVal  string
+		checker func(t *testing.T, cfg *EnvConfig)
```
Comment (pugnack) [inline env_time_test.go]: checker излишен, просто сделать want *EnvConfig

File: env_time_test.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+				if !cfg.TimeValue.Equal(expected) {
+					t.Errorf("expected %v, got %v", expected, cfg.TimeValue)
+				}
+			},
+		},
+	}
+
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			_ = os.Unsetenv("INT_VALUE")
```
Comment (pugnack) [inline env_time_test.go]: хрупкий тест, т.к. нужно следить за тем, чтобы не забывать его обновлять тут, также я не уверен, что в этом есть глубокий смысл, разве на каждый тест environment не сбрасывается? и даже если не сбрасывается, то ожидаемую переменную все равно выставяешь ведь каждый раз при тесте

File: env_time_test.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		envVal  string
+		checker func(t *testing.T, cfg *EnvConfig)
+	}{
+		{
+			name:   "int from env",
+			envVar: "INT_VALUE",
+			envVal: "100",
+			checker: func(t *testing.T, cfg *EnvConfig) {
+				if cfg.IntValue != 100 {
+					t.Errorf("expected 100, got %d", cfg.IntValue)
```
Comment (pugnack) [inline env_time_test.go]: используй библиотеку testify/requrie, чтобы не мучаться и не писать эти проверки, а просто сделать require.Equal(t, expected, result)

File: env.go [hunk 7]
```diff
@@ -81,7 +85,25 @@ func (c *envConfig) Load(ctx context.Context, opts ...config.LoadOption) error {
...
+
 	switch value.Kind() {
+	// Processing of type structures time.Time
+	case reflect.Struct:
+		if value.Type() == reflect.TypeOf(time.Time{}) {
+			layouts := []string{time.RFC3339}
+			var parsed time.Time
+			var err error
+
+			for _, layout := range layouts {
```
Comment (pugnack) [inline env.go]: ну и зачем нужен for, если ты объявляешь хардкод-слайс из одного элемента?

File: env.go [hunk 8]
```diff
@@ -373,5 +428,57 @@ func NewConfig(opts ...config.Option) config.Config {
...
+}
+
+func (t timeTransformer) Transformer(typ reflect.Type) func(dst, src reflect.Value) error {
+	// Processing of type structures time.Time
+	if typ == reflect.TypeOf(time.Time{}) {
+		return func(dst, src reflect.Value) error {
+			if !dst.CanSet() {
+				return nil
+			}
+			srcTime := src.Interface().(time.Time)
```
Comment (pugnack) [inline env.go]: зачем заниматься приведением к типу, если у тебя есть возможность через рефлексию проверять отношение к типу

File: env.go [hunk 9]
```diff
@@ -373,5 +428,57 @@ func NewConfig(opts ...config.Option) config.Config {
 	if len(options.StructTag) == 0 {
 		options.StructTag = DefaultStructTag
 	}
+
 	return &envConfig{opts: options}
 }
+
+type timeTransformer struct {
+	override bool
```
Comment (pugnack) [inline env.go:410]: а для чего нужен override?

File: env.go [hunk 10]
```diff
@@ -373,5 +428,57 @@ func NewConfig(opts ...config.Option) config.Config {
...
+	// Processing of type structures *time.Time
+	if typ == reflect.TypeOf((*time.Time)(nil)) {
+		return func(dst, src reflect.Value) error {
+			if !dst.CanSet() {
+				return nil
+			}
+			if src.IsNil() {
+				return nil
+			}
+			srcTime := src.Elem().Interface().(time.Time)
```
Comment (pugnack) [inline env.go]: аналогично, явное приведение типа здесь не нужно, т.к. ты работаешь с рефлексией и есть возможно пользоваться ее API

File: env.go [hunk 11]
```diff
@@ -274,6 +307,28 @@ func fillValues(ctx context.Context, valueOf reflect.Value, structTag string) er
 		}
 		switch value.Kind() {
 		case reflect.Struct:
+			if value.Type() == reflect.TypeOf(time.Time{}) {
+				tags, ok := field.Tag.Lookup(structTag)
```
Comment (pugnack) [inline env.go]: вижу кусок, который дублирует уже имеющийся код, вынеси в отдельную функцию

File: env.go [hunk 12]
```diff
@@ -274,6 +307,28 @@ func fillValues(ctx context.Context, valueOf reflect.Value, structTag string) er
 		}
 		switch value.Kind() {
```
Comment (pugnack) [inline env.go:301]: 1. IDE подсказывает, что секцию default стоит добавить с текущей логикой, котрая после switch
2. также поищи конструкции if len(field.PkgPath) != 0 — это по сути можно заменить на использование метода field.IsExported() посмотри реализацию, там то же поле проверяется, но выглядит очевиднее, только не перепутай семантику (!= и =)

File: go.mod [hunk 13]
```diff
@@ -9,9 +9,15 @@ require (
 	go.unistack.org/micro/v3 v3.11.14
 )
 
+require (
```
Comment (pugnack) [inline go.mod:12]: Нужен go mod tidy (в go.sum будет дифф)

File: env_test.go [hunk 14]
```diff
@@ -238,3 +468,36 @@ func TestLoadMultiple(t *testing.T) {
 		}
 	}
 }
+
+func TestEnv_ErrorHandling(t *testing.T) {
```
Comment (pugnack) [inline env_test.go]: Сомнительный тест, если уже пистать такое, то покрывать все supported types и валидировать, что все будут падать с ошибкой, либо убрать этот тест

File: env.go [hunk 15]
```diff
@@ -7,8 +7,10 @@ import (
 	"reflect"
 	"strconv"
 	"strings"
+	"time"
 
 	"dario.cat/mergo"
+
```
Comment (pugnack) [inline env.go]: пробел не нужен, т.к. все зависимости для текущего пакета, в том числе micro (это отдельная репа) — внешние

File: env.go [hunk 16]
```diff
@@ -157,6 +170,16 @@ func fillValue(ctx context.Context, value reflect.Value, val string) error {
 		}
 		value.Set(reflect.ValueOf(int32(v)))
 	case reflect.Int64:
+		t := value.Type()
+		// Processing of type structures time.Duration
```
Comment (pugnack) [inline env.go]: Некорректный коммент, time.Duration — не структура, можно убрать, ибо и так ясно, что просходит анализ time.Duration

File: env.go [hunk 17]
```diff
@@ -157,6 +170,16 @@ func fillValue(ctx context.Context, value reflect.Value, val string) error {
...
+		if t == reflect.TypeOf(time.Duration(0)) {
+			d, err := time.ParseDuration(val)
+			if err != nil {
+				return err
```
Comment (pugnack) [inline env.go]: добавь fmt.Errorf() с описанием по аналогии как в time.Time

File: env.go [hunk 18]
```diff
@@ -373,5 +393,60 @@ func NewConfig(opts ...config.Option) config.Config {
...
+
+			dst.Set(src)
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func getEnvByTag(field reflect.StructField, structTag string) ([]string, bool) {
```
Comment (pugnack) [inline env.go]: Ты изменила поведение, т.е. возвращаешь весь массив, а после бежишь по нему и пытаешься вызывать fillValues() для всех значений. Ранее брался последний элемент массива, если я не ошибаюсь, обрати на это внимание. То как сейчас — лишняя работа, т.к. ты просто перезаписываешь значения из полученного массива в одно и тоже поле  — не очень понятно зачем.

File: env_test.go [hunk 19]
```diff
@@ -19,6 +21,234 @@ type Config struct {
 	MapIntValue    map[string]int    `env:"MAP_INT"`
 }
 
+func TestEnv_SupportedTypes(t *testing.T) {
```
Comment (pugnack) [inline env_test.go]: В transformer явно видно, что ты работаешь с override, также я задавал уточняющий вопрос на который ты не ответила, но он был не просто так. Тест ты на этот кейс ты не описала, видимо была уверена, что оно работает. А зря, вот пару тестов, один из них упадет.

```
func TestEnv_TimeType_Override(t *testing.T) {
	type Config struct {
		TimeValue time.Time `env:"TIME"`
	}

	tests := []struct {
		name   string
		cfg    *Config
		envVar string
		envVal string
		want   *Config
	}{
		{
			name:   "init value is empty",
			cfg:    &Config{},
			envVar: "TIME",
			envVal: "2025-08-28T15:04:05Z",
			want: &Config{
				TimeValue: time.Date(2025, 8, 28, 15, 4, 5, 0, time.UTC),
			},
		},
		{
			name: "init value is not empty",
			cfg: &Config{
				TimeValue: time.Date(2025, 5, 25, 15, 5, 5, 5, time.UTC),
			},
			envVar: "TIME",
			envVal: "2025-08-28T15:04:05Z",
			want: &Config{
				TimeValue: time.Date(2025, 8, 28, 15, 4, 5, 0, time.UTC),
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			require.NoError(t, os.Setenv(tt.envVar, tt.envVal))
			defer os.Unsetenv(tt.envVar)

			cfg := NewConfig(config.Struct(tt.cfg))
			require.NoError(t, cfg.Init())
			require.NoError(t, cfg.Load(context.Background(), config.LoadOverride(true)))
			require.Equal(t, tt.want, tt.cfg)
		})
	}
}

func TestEnv_TimePointerType_Override(t *testing.T) {
	type Config struct {
		TimeValue *time.Time `env:"TIME"`
	}

	tests := []struct {
		name   string
		cfg    func() *Config
		envVar string
		envVal string
		want   func() *Config
	}{
		{
			name:   "init value is empty",
			cfg:    func() *Config { return &Config{} },
			envVar: "TIME",
			envVal: "2025-08-28T15:04:05Z",
			want: func() *Config {
				timeValue := time.Date(2025, 8, 28, 15, 4, 5, 0, time.UTC)
				return &Config{
					TimeValue: &timeValue,
				}
			},
		},
		{
			name: "init value is not empty",
			cfg: func() *Config {
				timeValue := time.Date(2025, 5, 25, 15, 5, 5, 5, time.UTC)
				return &Config{
					TimeValue: &timeValue,
				}
			},
			envVar: "TIME",
			envVal: "2025-08-28T15:04:05Z",
			want: func() *Config {
				timeValue := time.Date(2025, 8, 28, 15, 4, 5, 0, time.UTC)
				return &Config{
					TimeValue: &timeValue,
				}
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			require.NoError(t, os.Setenv(tt.envVar, tt.envVal))
			defer os.Unsetenv(tt.envVar)

			cfgData := tt.cfg()
			cfg := NewConfig(config.Struct(cfgData))
			require.NoError(t, cfg.Init())
			require.NoError(t, cfg.Load(context.Background(), config.LoadOverride(true)))
			require.Equal(t, tt.want(), cfgData)
		})
	}
}
```

File: env.go [hunk 20]
```diff
@@ -282,40 +318,24 @@ func fillValues(ctx context.Context, valueOf reflect.Value, structTag string) er
 		case reflect.Ptr:
 			if value.IsNil() {
 				if value.Type().Elem().Kind() != reflect.Struct {
-					// nil pointer to a non-struct: leave it alone
-					break
```
Comment (pugnack) [inline env.go]: Старайся делать минимальные  необходимые изменения, чтобы уменьшить кол-во шума в PR-ке, если речь не о рефакторинге, в данном случае просто сдвинулся коммент.

Это лишняя нагрузка на ревьюера.

File: env.go [hunk 21]
```diff
@@ -248,6 +269,20 @@ func (c *envConfig) setValues(ctx context.Context, valueOf reflect.Value) error
...
+	if !ok {
+		return "", false
+	}
+	var val string
+	for _, tag := range strings.Split(tags, ",") {
+		if v, ok := os.LookupEnv(tag); ok {
+			val = v
+		}
+	}
+	return val, len(val) > 0
```
Comment (pugnack) [inline env.go]: Теперь семантика корректная. 

Поправь len(val) > 0 => val != "", обычно так строки проверяют на пустоту (более читаемый подход)

**Comment by pugnack** [comment]
approved 👍
